cli/internal/cacheitem/cacheitem.go

```diff
@@ -0,0 +1,80 @@
+// Package cacheitem is an abstraction over the creation and restoration of a cache
+package cacheitem
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"crypto/sha512"
+	"errors"
+	"hash"
+	"io"
+	"os"
+
+	"github.com/vercel/turborepo/cli/internal/turbopath"
+)
+
+var (
+	errMissingSymlinkTarget = errors.New("symlink restoration is delayed")
+	errCycleDetected        = errors.New("links in the cache are cyclic")
+	errTraversal            = errors.New("tar attempts to write outside of directory")
+	errNameMalformed        = errors.New("file name is malformed")
+	errNameWindowsUnsafe    = errors.New("file name is not Windows-safe")
+	errUnsupportedFileType  = errors.New("attempted to restore unsupported file type")
+)
+
+// CacheItem is a `tar` utility with a little bit extra.
+type CacheItem struct {
+	// Path is the location on disk for the CacheItem.
+	Path turbopath.AbsoluteSystemPath
+	// Anchor is the position on disk at which the CacheItem will be restored.
+	Anchor turbopath.AbsoluteSystemPath
+
+	// For creation.
+	sha    hash.Hash
+	tw     *tar.Writer
+	gzw    *gzip.Writer
+	handle *os.File
+
+	// sticky error.
+	err error
+}
+
+// Close any open pipes
+func (ci *CacheItem) Close() error {
+	if ci.tw != nil {
+		if err := ci.tw.Close(); err != nil {
+			return err
+		}
+	}
+
+	if ci.gzw != nil {
+		if err := ci.gzw.Close(); err != nil {
+			return err
+		}
+	}
+
+	if ci.handle != nil {
```

Comment:
It can if you don't use `Open` or `Create`; but really I'm not sure what this will look like once I have "accepts streams" in the API. I'll hold on to it and remove it in the next PR if I find it unnecessary.